python/glow/gwas/approx_firth.py

```diff
@@ -0,0 +1,197 @@
+from nptyping import Float, NDArray
+from scipy import stats
+from statsmodels.base.model import Model
+import statsmodels.api as sm
+
+
+@dataclass
+class LogLikelihood:
+    pi: NDArray[(Any, ), Float]  # n_samples
+    I: NDArray[(Any, Any), Float]  # Fisher information matrix
+    deviance: Float  # -2 * penalized log likelihood
+
+
+@dataclass
+class FirthFit:
+    beta: NDArray[(Any, ), Float]  # n_covariates for null fit, 1 for SNP fit
+    log_likelihood: LogLikelihood
+
+
+@dataclass
+class FirthStatistics:
+    effect: Float
+    stderror: Float
+    tvalue: Float
+    pvalue: Float
+
+
+@typechecked
+def _calculate_log_likelihood(beta: NDArray[(Any, ), Float], model: Model) -> LogLikelihood:
+    pi = model.predict(beta)
+    I = -model.hessian(beta)
+    unpenalized_log_likelihood = model.loglike(beta)
+    _, log_abs_det = np.linalg.slogdet(I)
+    penalty = 0.5 * log_abs_det
+    deviance = -2 * (unpenalized_log_likelihood + penalty)
+    return LogLikelihood(pi, I, deviance)
+
+
+@typechecked
+def _fit_firth(beta_init: NDArray[(Any, ), Float],
+               X: NDArray[(Any, Any), Float],
+               y: NDArray[(Any, ), Float],
+               offset: NDArray[(Any, ), Float],
+               convergence_limit: float = 1e-5,
+               deviance_tolerance: float = 1e-6,
+               max_iter: int = 250,
+               max_step_size: int = 5,
+               max_half_steps: int = 25) -> Optional[FirthFit]:
+    '''
+    Firth’s bias-reduced penalized-likelihood logistic regression, based on the regenie implementation:
+    https://www.biorxiv.org/content/10.1101/2020.06.19.162354v2
+
+    Args:
+        beta_init : Initial beta values
+        X : Independent variable (covariate for null fit, genotype for SNP fit)
+        y : Dependent variable (phenotype)
+        offset : Phenotype offset only for null fit, phenotype + covariate effects for SNP fit
+        convergence_limit : Convergence is reached if all entries of the penalized score have smaller magnitude
+        deviance_tolerance : Non-inferiority margin when halving step size (default from regenie)
+        max_iter : Maximum number of Firth iterations (default from regenie)
+        max_step_size : Maximum step size during a Firth iteration (default from regenie)
+        max_half_steps : Maximum number of half-steps during a Firth iteration (default from regenie)
+
+    Returns:
+        None if the fit failed. Otherwise, a FirthFit object containing the fit information.
+    '''
+
+    n_iter = 0
+    beta = beta_init.copy()
+    model = sm.GLM(y, X, family=sm.families.Binomial(), offset=offset, missing='ignore')
+    log_likelihood = _calculate_log_likelihood(beta, model)
+
+    while n_iter < max_iter:
+        invI = np.linalg.pinv(log_likelihood.I)
+
+        # build hat matrix
+        rootG = np.sqrt(log_likelihood.pi * (1 - log_likelihood.pi))
+        rootG_X = np.expand_dims(rootG, 1) * X  # equivalent to np.diagflat(pi * (1 - pi)) @ X
```

Comment:
nits: I think it's more idiomatic (and common in our codebase) to use `rootG[:, None]`. Also, the comment doesn't include the `sqrt`.